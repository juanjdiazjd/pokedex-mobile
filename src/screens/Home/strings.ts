import LocalizedStrings from "react-native-localization";

export const  strings = new LocalizedStrings({
    es:{home:{
        infoTitle: 'Pokedex',
        infoSubtitle: 'All pokemons',
    }}
})

