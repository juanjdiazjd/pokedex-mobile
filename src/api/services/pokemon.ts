import { ApiResponse } from "apisauce";
import Config from "react-native-config";
import { Pokemon, PokemonList } from "../../types/Home/pokemon";

import buildApi from "../api";
import { URLS } from "./config";
const { pokemon } = URLS;
const LIMIT = 20;
const createApi = () => {
  // @ts-ignore
  const { get } = buildApi({ baseURL: Config.POKEAPI_URL });
  return {
    fetchAllPokemons: ({ pageParam = 0 }): Promise<PokemonList> =>
      get(`${pokemon.list}?limit=${LIMIT}&offset=${pageParam}`).then((res) => {
        return res.data as PokemonList;
      }),
    pokemonDetail: (
      data: any,
      pokemonId: string,
      config = {}
    ): Promise<ApiResponse<Pokemon>> =>
      get(`${pokemon.detail}/${pokemonId}`, data, config),
  };
};

export default createApi;
