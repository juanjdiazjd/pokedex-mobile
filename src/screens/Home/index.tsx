import * as React from "react";
import { FlatList, ListRenderItem, StatusBar } from "react-native";
import Header from "../../components/Header";
import {
  getOffsetFromUrl,
  pokemonItemExtractorKey,
  RootStackParamList,
} from "./utils";
import { NativeStackScreenProps } from "@react-navigation/native-stack";
import { WrapperView } from "../../components/Wrappers/SafeAreaWrapper";
import { ContentView } from "../../components/Wrappers/ContentView";
import { useInfiniteQuery } from "react-query";

import { strings } from "./strings";
import { TextType } from "../../components/Text/TextView";
import { NamedAPIResource } from "../../types/Home/pokemon";
import { Card } from "../../components/Pokemon/Card";
import { PokemonLogo } from "../../components/Pokemon/styledComponents";
import pokemonApi from "../../api";
import { MarginSize } from "../../types/sizes";

const { pokemonServices: pokemonAPI } = pokemonApi;
const NUM_COLUMNS = 2;

const HomeScreen = ({
  navigation,
  route,
}: NativeStackScreenProps<RootStackParamList, "Home">) => {
  const { isLoading, data, hasNextPage, fetchNextPage, isFetchingNextPage } =
    useInfiniteQuery("pokemons", pokemonAPI.fetchAllPokemons, {
      getNextPageParam: (lastPage) => {
        if (lastPage.next !== null) {
          return getOffsetFromUrl(lastPage.next);
        }
        return lastPage;
      },
    });

  const loadMore = () => {
    if (hasNextPage) {
      fetchNextPage();
    }
  };
  const onListItemPress = React.useCallback(
    (pokemon) => {
      return navigation.navigate("PokemonDetail", pokemon);
    },
    [navigation]
  );

  const renderItem: ListRenderItem<NamedAPIResource> = ({ item }) => (
    <Card onPress={onListItemPress} name={item.name} url={item.url} />
  );

  return (
    <WrapperView>
      <ContentView fullWidth={true}>
        <StatusBar barStyle="dark-content" />
        <Header
          title={strings.home.infoTitle}
          subtitle={strings.home.infoSubtitle}
          withoutBackButton={false}
          paddingHeading={MarginSize.mediumPadding}
          textTypeTitle={TextType.bigBoldMonserrat}
          textTypeSubtitle={TextType.verySmallBold}
          secondaryComponent={() => (
            <PokemonLogo
              resizeMode="cover"
              source={
                isFetchingNextPage
                  ? require("../../assets/pokeball-logo.png")
                  : require("../../assets/pokeball-grey.png")
              }
            />
          )}
        />
        <FlatList
          data={data?.pages.map((page) => page.results).flat()}
          numColumns={NUM_COLUMNS}
          contentContainerStyle={{ alignSelf: "center" }}
          keyExtractor={pokemonItemExtractorKey}
          renderItem={renderItem}
          onEndReached={loadMore}
          onEndReachedThreshold={0.3}
        />
      </ContentView>
    </WrapperView>
  );
};

export default HomeScreen;
