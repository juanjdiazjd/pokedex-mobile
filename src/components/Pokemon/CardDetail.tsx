import * as React from "react";
import { ScrollView, useWindowDimensions, View } from "react-native";
import styled from "styled-components";
import theme from "../../theme";
import { TextType, TextView } from "../Text/TextView";
import { convertPokemonIdToText } from "./utils/index";
import { Pokemon } from "../../types/Home/pokemon";
import { PokemonDetailSprite } from "./styledComponents";
import { Button, List } from "react-native-paper";

interface PokemonDetailCardProps {
  pokemonDetail: Partial<Pokemon>;
}

export const Wrapper = styled(View)`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
`;
export const AdditionalInfoWrapper = styled(View)<{ width: number }>`
  border-radius: 20px;
  width: ${({ width }) => width}px;
  height: 460px;
  background-color: white;
`;
export const PokemonTypeWrapper = styled(View)`
  margin: 2px;
  border-radius: 10px;
  align-self: center;
  background-color: rgba(0, 0, 0, 0.17);
  padding: 10px;
`;

const Row = styled(View)`
  flex-direction: row;
  padding: 20px;
  align-content: space-between;
`;

export const CardDetail: React.FunctionComponent<PokemonDetailCardProps> = ({
  pokemonDetail,
}) => {
  const windowDimensions = useWindowDimensions();

  return (
    <View>
      <Row>
        <TextView
          id={"pokemon_name"}
          text={`${pokemonDetail?.name?.toLowerCase()} - ${convertPokemonIdToText(
            pokemonDetail?.id || 0
          )}`}
          type={TextType.veryBigBoldMonserrat}
          style={{ textTransform: "capitalize" }}
          textAlign={"center"}
          colorText={theme.colors.white}
        ></TextView>
      </Row>
      <Row style={{ justifyContent: "center" }}>
        {pokemonDetail?.types?.map((type, index) => (
          <PokemonTypeWrapper key={index}>
            <TextView
              id={`pokemon_type_${index}`}
              text={type.type.name}
              type={TextType.bigBoldMonserrat}
              colorText={theme.colors.white}
              textAlign={"center"}
            />
          </PokemonTypeWrapper>
        ))}
      </Row>

      <PokemonDetailSprite
        source={{
          uri: pokemonDetail?.sprites?.other["official-artwork"].front_default,
        }}
      />
      <AdditionalInfoWrapper width={windowDimensions.width}>
        <Button
          icon="arrow-up"
          mode='text'
          color={theme.colors.black}
          onPress={() => console.log("Pressed")}
        >
        </Button>
        <List.Section style={{}}>
          <ScrollView>
            <List.Subheader>Abilities: </List.Subheader>
            {pokemonDetail.abilities?.map((data, index) => (
              <List.Item key={index.toString()} title={data.ability.name} />
            ))}
          </ScrollView>
        </List.Section>
      </AdditionalInfoWrapper>
    </View>
  );
};
