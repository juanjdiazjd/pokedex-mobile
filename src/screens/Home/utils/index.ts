import { getPokemonIdFromUrl } from "../../../components/Pokemon/utils";
import { NamedAPIResource, Pokemon } from "../../../types/Home/pokemon";

export const icons: Record<string, string> = {
  Home: "pokemon-go",
  PokemonDetail: "detail",
};

export type RootStackParamList = {
  Home: undefined;
  PokemonDetail: Pokemon;
};

export const getOffsetFromUrl = (url: any | string) => {
  if (url !== null && url !== undefined) {
    return url.split("/")[5].match("=(.*)&")[1];
  }
};
export const pokemonItemExtractorKey = (
  item: NamedAPIResource,
  index: number
) => {
  return getPokemonIdFromUrl(item.url).toString();
};
