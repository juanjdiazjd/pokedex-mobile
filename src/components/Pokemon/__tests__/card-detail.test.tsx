import React from "react";
import { CardDetail } from "../CardDetail";
import renderer from 'react-test-renderer';

import { Pokemon } from "../../../types/Home/pokemon";
jest.mock("axios");

const MOCK_POKEMON: Partial<Pokemon> = {
  id: 6,
  name: "charizard",
  sprites: {
    front_default:
      "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/home/6.png",
    front_female: "",
    front_shiny:
      "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/home/shiny/6.png",
    front_shiny_female: "",
    back_default:
      "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/home/6.png",
    back_female: "",
    back_shiny:
      "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/home/shiny/6.png",
    back_shiny_female: "",
    other: {
      dream_world: {
        front_default:
          "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/6.svg",
        front_female: "",
      },
      home: {
        front_default:
          "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/home/6.png",
        front_female: "",
        front_shiny:
          "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/home/shiny/6.png",
        front_shiny_female: "",
      },
      "official-artwork": {
        front_default:
          "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/6.png",
      },
    },
  },
  types: [
    {
      slot: 1,
      type: {
        name: "fire",
        url: "https://pokeapi.co/api/v2/type/10/",
      },
    },
    {
      slot: 2,
      type: {
        name: "flying",
        url: "https://pokeapi.co/api/v2/type/3/",
      },
    },
  ],
  abilities: [
    {
      ability: {
        name: "blaze",
        url: "https://pokeapi.co/api/v2/ability/66/",
      },
      is_hidden: false,
      slot: 1,
    },
    {
      ability: {
        name: "solar-power",
        url: "https://pokeapi.co/api/v2/ability/94/",
      },
      is_hidden: true,
      slot: 3,
    },
  ],
};

it("Should pokemon card detail component", () => {
  const tree = renderer
    .create(<CardDetail pokemonDetail={MOCK_POKEMON} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
