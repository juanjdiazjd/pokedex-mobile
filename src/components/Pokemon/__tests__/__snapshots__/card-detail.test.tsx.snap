// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should pokemon card detail component 1`] = `
<View>
  <View
    className="sc-iCfMLu dmXPVK"
  >
    <Text
      className="sc-bdvvtL dPVMfN"
      colorText="#FFFFFF"
      mode="light"
      style={
        Object {
          "textTransform": "capitalize",
        }
      }
      textAlign="center"
      values={
        Object {
          "color": "#FFFFFF",
          "fontSize": "28px",
          "fontWeight": "800",
          "lineHeight": "28px",
          "type": "veryBigBoldMonserrat",
          "variant": "Montserrat-SemiBold",
        }
      }
    >
      charizard - #006
    </Text>
  </View>
  <View
    className="sc-iCfMLu dmXPVK"
    style={
      Object {
        "justifyContent": "center",
      }
    }
  >
    <View
      className="sc-gKclnd bJeKtN"
    >
      <Text
        className="sc-bdvvtL gsWCfe"
        colorText="#FFFFFF"
        mode="light"
        textAlign="center"
        values={
          Object {
            "color": "#FFFFFF",
            "fontSize": "18px",
            "fontWeight": "800",
            "lineHeight": "28px",
            "type": "bigBoldMonserrat",
            "variant": "Montserrat-SemiBold",
          }
        }
      >
        fire
      </Text>
    </View>
    <View
      className="sc-gKclnd bJeKtN"
    >
      <Text
        className="sc-bdvvtL gsWCfe"
        colorText="#FFFFFF"
        mode="light"
        textAlign="center"
        values={
          Object {
            "color": "#FFFFFF",
            "fontSize": "18px",
            "fontWeight": "800",
            "lineHeight": "28px",
            "type": "bigBoldMonserrat",
            "variant": "Montserrat-SemiBold",
          }
        }
      >
        flying
      </Text>
    </View>
  </View>
  <Image
    className="sc-hKwDye dwjBKK"
    source={
      Object {
        "uri": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/6.png",
      }
    }
  />
  <View
    className="sc-jRQBWg hgamtG"
    width={750}
  >
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "transparent",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "elevation": 0,
          "minWidth": 64,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 4,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Object {
                "marginLeft": 12,
                "marginRight": -4,
              }
            }
          >
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontSize": 16,
                  },
                  Array [
                    Object {
                      "lineHeight": 16,
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              Û∞Åù
            </Text>
          </View>
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  Object {
                    "textTransform": "uppercase",
                  },
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Object {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "marginVertical": 8,
          },
          Object {},
        ]
      }
      theme={
        Object {
          "animation": Object {
            "scale": 1,
          },
          "colors": Object {
            "accent": "#03dac4",
            "backdrop": "rgba(0, 0, 0, 0.5)",
            "background": "#f6f6f6",
            "disabled": "rgba(0, 0, 0, 0.26)",
            "error": "#B00020",
            "notification": "#f50057",
            "onSurface": "#000000",
            "placeholder": "rgba(0, 0, 0, 0.54)",
            "primary": "#6200ee",
            "surface": "#ffffff",
            "text": "#000000",
          },
          "dark": false,
          "fonts": Object {
            "light": Object {
              "fontFamily": "System",
              "fontWeight": "300",
            },
            "medium": Object {
              "fontFamily": "System",
              "fontWeight": "500",
            },
            "regular": Object {
              "fontFamily": "System",
              "fontWeight": "400",
            },
            "thin": Object {
              "fontFamily": "System",
              "fontWeight": "100",
            },
          },
          "roundness": 4,
        }
      }
    >
      <RCTScrollView>
        <View>
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "paddingHorizontal": 16,
                    "paddingVertical": 13,
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.54)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            Abilities: 
          </Text>
          <View
            accessibilityState={
              Object {
                "disabled": true,
              }
            }
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                false,
                Array [
                  Object {
                    "padding": 8,
                  },
                  undefined,
                ],
              ]
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "marginVertical": 6,
                      "paddingLeft": 8,
                    },
                    Object {
                      "flex": 1,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Array [
                        Object {
                          "fontSize": 16,
                        },
                        Object {
                          "color": "rgba(0, 0, 0, 0.87)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  blaze
                </Text>
              </View>
            </View>
          </View>
          <View
            accessibilityState={
              Object {
                "disabled": true,
              }
            }
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                false,
                Array [
                  Object {
                    "padding": 8,
                  },
                  undefined,
                ],
              ]
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "marginVertical": 6,
                      "paddingLeft": 8,
                    },
                    Object {
                      "flex": 1,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Array [
                        Object {
                          "fontSize": 16,
                        },
                        Object {
                          "color": "rgba(0, 0, 0, 0.87)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  solar-power
                </Text>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
